src/main/scala/org/broadinstitute/hail/driver/FilterAlleles.scala

```diff
@@ -126,29 +144,58 @@ object FilterAlleles extends Command {
 
       def downcodeGenotype(g: Genotype): Genotype = {
         val px = g.px.map(downcodePx)
-        val gq = px.map(Genotype.gqFromPL)
-        Genotype(g.gt.map(downcodeGt),
-          g.ad.map(downcodeAd),
-          g.dp,
-          gq,
-          px,
-          g.fakeRef,
-          g.isDosage
+        g.copy(gt = g.gt.map(downcodeGt),
+          ad = g.ad.map(downcodeAd),
+          gq = px.map(Genotype.gqFromPL),
+          px = px
+        )
+      }
+
+      def truncatePx(px: Array[Int]): Array[Int] = {
+        val (newPx,minP) = px.zipWithIndex
+          .filter({
+            case (p, i) =>
+              val gTPair = Genotype.gtPair(i)
+              (gTPair.j == 0 || oldToNew(gTPair.j) != 0) && (gTPair.k ==0 || oldToNew(gTPair.k) != 0)
+          })
+          .foldLeft((Array.fill(triangle(newCount))(0),Int.MaxValue))({
+            case((newPx,minP),(p,i)) =>
+              newPx(downcodeGt(i)) = p
+              (newPx,min(p,minP))
+          })
+
+        newPx.map(_ - minP)
+      }
+
+      def truncateGenotype(g: Genotype) : Genotype = {
+        val px  = g.px.map(truncatePx)
+        g.copy(
+          gt = px.map(_.zipWithIndex.min._2),
+          ad = g.ad.map(_.zipWithIndex.filter({case(d,i) => i ==0 || oldToNew(i) != 0}).map(_._1)),
+          gq = px.map(Genotype.gqFromPL),
+          px = px
         )
       }
 
-      gs.map(downcodeGenotype)
+      gs.map({
+        g =>
+          val newG = if(downcode) downcodeGenotype(g) else truncateGenotype(g)
+          if(filterAlteredGenotypes && newG.gt != g.gt)
+            newG.copy(gt = None)
+          else
+            newG
+      })
     }
 
     def updateOrFilterRow(v: Variant, va: Annotation, gs: Iterable[Genotype]): Option[(Variant, (Annotation, Iterable[Genotype]))] =
       filterAllelesInVariant(v, va).map { case (newV, newToOld, oldToNew) =>
-        val newVa = updateAnnotation(v, va, newToOld)
+        val newVa = if(updateAnn) updateAnnotation(v, va, newToOld) else va
         val newGs = updateGenotypes(gs, oldToNew, newToOld.length)
         (newV, (newVa, newGs))
       }
 
     val newVds = state.vds
-      .flatMapVariants { case (v, va, gs) => updateOrFilterRow(v, va, gs) }
+      .flatMapVariants { case (v, va, gs) => updateOrFilterRow(v, va, gs) } // should this be map -> filter ?
```

Comment:
it would have to be
```scala
state.vds
  .mapVariants { case (v, va, gs) => updateOrFilterRow(v, va, gs) }
  .filter { x => x.isPresent }
  .map { x => x.get }
```
which is a bit more verbose and feels less safe-y because we call `x.get`.